### PR TITLE
feat: option to disallow project initialization

### PR DESCRIPTION
**Pull request by nunomiguel22**
This provides an option to disable initializing projects when a solution is already loaded. Reason for this is the project structure described in the Readme file. Currently it prevents me from jumping from a project that is outside the solution directory to projects inside it.

By default it will not alter the behaviour but will give an option to handle this scenario, which has been my only hiccup in an otherwise great experience. If you need further explanation on this let me know.

### CODE REVIEW

**Comment by seblj** [comment]
Hmm, I don't think I fully understand. Could you try to explain a bit more in detail what you want?

**Comment by nunomiguel22** [comment]
Sure, here's a simple practical example.
Step 1 - Create a solution with a single project, add a single file to it (Class1)
Step 2 - Create a project with a single file (Class2) outside of the root directory of the solution and reference it in the solution. So that at the end you have something like this:
image
Step 3 - Call a method from Class1 in Class2 
image
Step 4 - Build the solution

Now if you open Class 1 it will correctly load dotnet-test.sln, however if you go to Class2 it will initialize an lsp server for ConsoleApp1.csproj as a project, since it is outside of the root directory of dotnet-test.sln:
image

So if you 'gd' the method you called from Class1 it will jump to the decompiled assembly:
image
image

And this is the incorrect behaviour for this use case, when you initialize dotnet-test.sln it will already consider ConsoleApp1.csproj. So if you stop a lsp server from being started when you open Class2, everything will behave as you would expect.
Set allow_csproj_for_sln_swap to false, go to Class1 and let dotnet-test.sln be initialized, go to Class2 and if you gd the same method it will jump to the file instead of the decompiled assembly.
image

Pretty much all solutions i work with will reference projects outside of the solution directory making it impossible to jump from those "external" projects into the projects in the solution directory.

The solution provided here is my current workaround for this, to have a way to stop the lsp server from being initiated in projects when i already have a solution loaded. Perhaps there is a better way to do this but it works pretty well for me right now while still being able to load projects if i want. If this is not clear enough i can record the issue.

**Comment by seblj** [comment]
Okay thanks! I think I understand now. I think I need a couple of days to think about this, because ideally I would like this to just work with no further options added.

A while ago I tried looking into finding out which projects are part of a solution, but I couldn't find a good way.

I think, if there is a good way to get all projects related to a solution, I don't think we would need this option.

Do you know of a way to do that other than just parse the solution file manually?

**Comment by seblj** [comment]
Does this fix your problem? https://github.com/seblj/roslyn.nvim/pull/83

**Comment by nunomiguel22** [comment]
Yes, i tested your solution and it worked great. Although I do think having an option to lock in the solution you want and never change it unless the user wants(vscode/vs2002 style) could be a valuable option to have. This way if the plugin default behaviour is not what the user expects there would be a way to workaround it. This is just a suggestion, for my use case your solution seems completely fine.

**Comment by seblj** [comment]
Yeah, that I also think can be a pretty good option. I have been experimenting a little bit more with how it should attach to projects/solutions now, and I think I found an okay way (IMO).

It will change a bit for people, and I am not sure if everyone will like it, but I am also unsure if it's possible to satisfy everyone🤔 

I need to think about it a bit if I want to continue down the route I have experimented with. I might update the PR or create a new one, but I will probably be adding an option to lock in to a solution file as you say. I think that makes sense!

**Comment by seblj** [comment]
I was also kind of thinking about the same thing with a project🤔 What do you think about for example adding possible `csproj` files to the `Roslyn target` selection, and then having the "locking" also lock it to a project as well?

**Comment by nunomiguel22** [comment]
Yea there will never be a perfect solution for every case, that's why vscode/vs2022 don't even bother to try, they just force you to lock in one. I think this implementation is very flexible, if i want to initialize something as the project i just need to go to the project files and if i need to use it in the context of a solution i initialize the solution first then open the project files. It allows the user to make a deliberate choice.

Locking a solution or project would both be valuable options just as long as they are optional. Locking a project is a more niche use case(VS Code does not even allow it), so it depends on the amount of work it would take. The ability to lock a solution is far more important in my opinion.

**Comment by seblj** [comment]
Yeah okay that makes sense! I am now just considering locking it by default.

So that on first file open, it does all the regular stuff. Looks for a solution file, tries to guess it if there are multiple, and if it can't guess it, then you have to choose ofc. And then when it has loaded for the first time, it just always uses that until the user then does `:Roslyn target` again.

I actually created this for work, and I haven't really worked in a lot of projects with C#, so I really appreciate input from others on these kind of things. What do you think about making this the default option? It seems like a very safe bet?

**Comment by nunomiguel22** [comment]
I think it is ok to have it as a default option as that's what most IDEs already do but the drawback of that option is that you lose some agility that you gain now compared to other implementations. Most of the c# projects i work with will have atleast a Business solution and a Test solution. So if i were to use VS Code i would have to be switching constantly between the two if i want to use the same editor instance with LSP support.

Right now with the current implementation I can navigate between both solutions with LSP support instantly, it is really nice. Of there is some trade-offs, you will use more ram and in more complex setups you may not get the behaviour you want. Having the option to have the way you prefer would be the most important thing, as for the default option it is debatable. I would say locking by default probably makes more sense because new users coming from other IDEs will probably expect that behaviour.

**Comment by seblj** [comment]
Okay thanks for the info! I have implemented an alternative to how it works today in main, but I am still a bit unsure if I like it. As you say, there are always positives and negatives.

The PR I linked to is now updated with the new way.

I can try to describe it:

1. If there is already a selected target, and the lock option is one, then use that.
2. If we find both a solution file and a csproj file, then check if the project is a part of the solution.
  - If it is: use the solution
  - If it's not: start the project
3. If we only find a solution file, then use that
4. If we only find a project file, then use that
5. If we have a selected solution file, fallback to that.

I really appreciate the input, and I would love to get some feedback on this approach as well. If others see this, then feel free the chime in

# Positive:

I am now able to navigate to a standalone project and actually get the server working in the same instance.

This is currently not possible, as it would just use the solution file, since we don't verify that it is a part of the solution

# Negatives

It might use more resources and start up unnecessary things if the wrong file is opened by accident

I also don't know if the method I made to check if it the project is a part of the solution works 100% as expected, but that will of course be fixed if some bugs appear because of that

TLDR; It checks if the found project is part of the found solution, and starts the project if it's not. Instead of just always selecting the found solution

**Comment by nunomiguel22** [comment]
Seems good to me, I work with all kinds of C# projects daily, so if i find issues with it i will report it

**Comment by seblj** [comment]
I just merged the PR i have been working on that I think should fix your issue. I hope it works as expected for people.

I actually am pretty satisfied with the solution I found, so I hope others also will be happy with it😄

Let me know if you have any issues or concerns with it.

I decided to make the locking default to false btw
